coderd/agentapi/devcontainer_agent_test.go

```diff
@@ -0,0 +1,408 @@
+		})
+		resource := dbgen.WorkspaceResource(t, db, database.WorkspaceResource{
+			JobID: build.JobID,
+		})
+		agent := dbgen.WorkspaceAgent(t, db, database.WorkspaceAgent{
+			ResourceID: resource.ID,
+		})
+
+		return user, agent
+	}
+
+	newAgentAPI := func(t *testing.T, logger slog.Logger, db database.Store, clock quartz.Clock, user database.User, org database.Organization, agent database.WorkspaceAgent) *agentapi.DevContainerAgentAPI {
+		auth := rbac.NewStrictCachingAuthorizer(prometheus.NewRegistry())
+
+		accessControlStore := &atomic.Pointer[dbauthz.AccessControlStore]{}
+		var acs dbauthz.AccessControlStore = dbauthz.AGPLTemplateAccessControlStore{}
+		accessControlStore.Store(&acs)
+
+		return &agentapi.DevContainerAgentAPI{
+			OwnerID:        user.ID,
+			OrganizationID: org.ID,
+			AgentID:        agent.ID,
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Clock:    clock,
+			Database: dbauthz.New(db, auth, logger, accessControlStore),
+		}
+	}
+
+	t.Run("CreateDevContainerAgent", func(t *testing.T) {
+		t.Parallel()
+
+		tests := []struct {
+			name      string
+			agentName string
+			agentDir  string
+			agentArch string
+			agentOS   string
+			shouldErr bool
+		}{
+			{
+				name:      "Ok",
+				agentName: "some-child-agent",
+				agentDir:  "/workspaces/wibble",
+				agentArch: "amd64",
+				agentOS:   "linux",
+			},
+			{
+				name:      "NameWithUnderscore",
+				agentName: "some_child_agent",
+				agentDir:  "/workspaces/wibble",
+				agentArch: "amd64",
+				agentOS:   "linux",
+				shouldErr: true,
+			},
+			{
+				name:      "EmptyName",
+				agentName: "",
+				agentDir:  "/workspaces/wibble",
+				agentArch: "amd64",
+				agentOS:   "linux",
+				shouldErr: true,
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				t.Parallel()
+
+				log := testutil.Logger(t)
+				ctx := testutil.Context(t, testutil.WaitShort)
+				clock := quartz.NewMock(t)
+
+				db, org := newDatabaseWithOrg(t)
+				user, agent := newUserWithWorkspaceAgent(t, db, org)
+				api := newAgentAPI(t, log, db, clock, user, org, agent)
+
+				createResp, err := api.CreateDevContainerAgent(ctx, &proto.CreateDevContainerAgentRequest{
+					Name:            tt.agentName,
+					Directory:       tt.agentDir,
+					Architecture:    tt.agentArch,
+					OperatingSystem: tt.agentOS,
+				})
+				if tt.shouldErr {
+					require.Error(t, err)
+				} else {
+					require.NoError(t, err)
+
+					agentID, err := uuid.FromBytes(createResp.Id)
+					require.NoError(t, err)
+
+					agent, err := api.Database.GetWorkspaceAgentByID(dbauthz.AsSystemRestricted(ctx), agentID) //nolint:gocritic // this is a test.
+					require.NoError(t, err)
+
+					require.Equal(t, tt.agentName, agent.Name)
+					require.Equal(t, tt.agentDir, agent.Directory)
+					require.Equal(t, tt.agentArch, agent.Architecture)
+					require.Equal(t, tt.agentOS, agent.OperatingSystem)
```

Comment:
Suggest `assert.Equal` here so you get more information and not just the first failure.